### PR TITLE
Add meandepth to print template

### PR DESCRIPTION
**Pull request by jb2cool**
Adding additional metric meandepth to be available to the print template.

Reported-by: Rahul Swaminathan <rahul.swaminathan@gmail.com>
Signed-off-by: Jason Bramwell <jb2cool@gmail.com>

### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### CODE REVIEW

**Comment by atdotde** [comment]
Have you actually checked that with this you can use meandepth in a template?

BTW, get_depth_string returns the depth according to the unit preferences, so it is not necessarily metric.

**Comment by atdotde** [comment]
And while at it, also export the partial gas fractions (I would export those as percentages, i.e. get_o2(gasmix)/10 etc(,

**Comment by jb2cool** [comment]
> Have you actually checked that with this you can use meandepth in a template?
> 
> BTW, get_depth_string returns the depth according to the unit preferences, so it is not necessarily metric.

No, you are right, this hasn't actually be tested in anger. I'll pull this for now till i can test this fully. Just being a bit keen is all.

**Comment by dirkhh** [comment]
Thanks for doing this, Jason. I would not have closed this, you should just continue to work on it to get it to the point where it's tested and works.
I am super excited to see code contributions from you and will be happy to make time to help you get this to that point.

**Comment by jb2cool** [comment]
Just built this locally and i don't actually get the print option available to me under the File menu on my locally built version.

I think i need to ./subsurface/scripts/build.sh -desktop instead to get this, trying this now.

**Comment by dirkhh** [comment]
you need to make sure that you have QtWebkit installed and that you build against it.
Otherwise printing is turned off

**Comment by jb2cool** [comment]
Thanks, i see the instructions for getting that now, should have thought about grepping for print in the INSTALL file.

**Comment by dirkhh** [comment]
ok, I was able to build this and the code as submitted seems to work just fine...
from my edited template:
```
	<td class="fieldTitle">
		<h1> Max / mean depth </h1>
	</td>
	<td class="fieldData">
		<p> {{ dive.depth }} / {{ dive.meandepth }} </p>
	</td>
```
And the result:
Screen Shot 2022-03-03 at 11 35 56 AM
which looks right to me.
Which means that this pull request is on the right track. We should add this to the documentation, and we should add the individual gas values that were requested while we are at it - but this is looking great so far!

**Comment by dirkhh** [comment]
and yes, Robert is right - the commit message is wrong in claiming that this is metric :)

**Comment by jb2cool** [comment]
Ah that was maybe just poor wording on my part, i meant metric as in unit
of measure rather than metric in the imperial/metric system, that had never
even occurred to me. Perhaps i should have just called this a value or
string.

Still struggling getting Webkit installed here.

JB

**Comment by dirkhh** [comment]
> Ah that was maybe just poor wording on my part, i meant metric as in unit of measure rather than metric in the imperial/metric system, that had never even occurred to me. Perhaps i should have just called this a value or string.

agreed

> Still struggling getting Webkit installed here. JB

which OS are you on? Most of the Linux distros still offer it as a package. If you don't have access to that, then it's an incredible pain in the...

**Comment by jb2cool** [comment]
I think my Webkit is good now. I had started going down the local compiling route but now it've just picked it up as a package from the repo.

Am i now making a mess of things with the way I've amended things? happy for you to reject this entirely and then i'll trash and refork my repo where i can hopefully do everything cleanly in one commit.

**Comment by dirkhh** [comment]
> I think my Webkit is good now. I had started going down the local compiling route

OMG no, don't do that. There lies insanity. I have to do that for Windows and macOS and that is the sink of 90% of the time that goes into upgrading Qt versions. It is insanely frustrating to get this to work and the smallest changes can break things. I so wish we had never gone down the WebKit path...

> but now it've just picked it up as a package from the repo.

Good
 
> Am i now making a mess of things with the way I've amended things? happy for you to reject this entirely and then i'll trash and refork my repo where i can hopefully do everything cleanly in one commit.

It's a bit of a mess but that's really easy for me to clean up as I merge it. Don't worry about that part.

**Comment by jb2cool** [comment]
I'm struggling to get the gas fractions working with the printing templates.

The logical place appears to put this under cylinder objects (desktop-widgets/templatelayout.cpp line 489 to 505) but then i don't know how i call this from the print template. If i try putting this under the dive list (desktop-widgets/templatelayout.cpp lines 506 to 588) then the build process seems to fail for all the different syntax i've tried (probably doing something stupid). Am i being naive thinking that desktop-widgets/templatelayout.cpp is the only file i need to touch or should i have to do things to files like core/string-format.h & core/string-format.cpp too? As you might guess, i'm shooting pretty blind. I think with poking around that yes, i suspect anything in desktop-widgets/templatelayout.cpp probably does need to be mirrored in core/string-format.h & core/string-format.cpp.

Possibly another thing that isn't working where i think it should is that I think i should be able to call dive.cylinderList but i cant seem to make this do anything, this might be me misreading things though.

**Comment by dirkhh** [comment]
You need something like this:
```
                {% for cyl in cylinderObjects %}                                                            
                            <tr>
                                <td>{{cyl.description}}</td>
                                <td>{{cyl.gasMix}}</td>
                                <td>{{cyl.startPressure}}</td>
                                <td>{{cyl.endPressure}} </td>
                {% endfor %}
```
and instead of just offering `gasMix`, you'll also offer `cyl.he` and `cyl.o2`

From an implementation perspective, `get_o2(cylinder->gasmix) + 5) / 10` gives you the O2 percentage...

**Comment by jb2cool** [comment]
I'd actually set them up as cyl.gasHe, cyl.gasN2 and cyl.gasO2 but can
change them if you want. Thanks for that template example, that helped and
yes, using that my changes do indeed seem to work ok, will commit and push
that soon.

Can you explain to me the get_o2(cylinder->gasmix) + 5) / 10. I currently
have get_o2(cylinder->gasmix)/10 and that seems to work correctly.

Jason

**Comment by dirkhh** [comment]
> I'd actually set them up as cyl.gasHe, cyl.gasN2 and cyl.gasO2 but can change them if you want. 

no, that's perfectly fine

> Thanks for that template example, that helped and yes, using that my changes do indeed seem to work ok, will commit and push that soon.

nice :)

> Can you explain to me the `(get_o2(cylinder->gasmix) + 5) / 10`. I currently have `get_o2(cylinder->gasmix)/10` and that seems to work correctly.

Well, since we keep pressures in mbar we have permille here, so in order to get percent we divide by 10... and my code simply contains the correct rounding (for integer arithmetic).

**Comment by jb2cool** [comment]
How does this look? Seems to do the right things.

[print.pdf](https://github.com/subsurface/subsurface/files/8188025/print.pdf)

**Comment by dirkhh** [comment]
That is quite the dive - but the data look good!

File: Documentation/user-manual.html.git [hunk 1]
```diff
@@ -7661,6 +7661,10 @@ cellspacing="0" cellpadding="4">
 <td align="left" valign="top"><p class="table">(<strong>string</strong>) depth of the dive</p></td>
 </tr>
 <tr>
+<td align="left" valign="top"><p class="table">meandepth</p></td>
+<td align="left" valign="top"><p class="table">(<strong>string</strong>) mean depth of the dive</p></td>
+</tr>
+<tr>
```
Comment (dirkhh) [inline Documentation/user-manual.html.git:7667]: BTW: don't update the `.html.git` file - this is automatically created by my tools.

**Comment by dirkhh** [comment]
This reminds me that we should make sure to slowly remove all references to `Grantlee` from the manual (we don't actually use Grantlee anymore). We also should make sure that we document all the available variables - I'm not sure that we do right now... 

That said, neither of these are things that you need to do in order for this PR to get merged.

**Comment by jb2cool** [comment]
Fingers crossed this hopefully addresses everything.

**Comment by dirkhh** [comment]
this looks excellent.
I will rearrange the commits a little bit locally and push a cleaned-up version.
Thank you so much for working on this!

**Comment by jb2cool** [comment]
@dirkhh i appreciate your assistance in showing me how i can call my newly created variables and also the initial pointers from @atdotde to show me what needed to be done.

**Comment by dirkhh** [comment]
BTW: the DCO test is failing because you used `Signed=off-by` instead of `Signed-off-by`
I fixed that as well and will push this branch to master in a few moments.

**Comment by jb2cool** [comment]
Damn, sorry about that, dumb typo. That's me juggling things on different computers, working tired and typing in sub-optimal conditions (reaching around stuff to get to keyboard).

Hopefully my next contribution (where i can find something) will be a bit neater.

**Comment by dirkhh** [comment]
This took me a couple of minutes to fix. Well worth it and not a problem at all.

These days I am thrilled beyond believe when I see contributions - I'm looking forward to your next one.

**Pull request closed without merging**